File: js/WebSocketClient.js
"use strict"

import { EventManager } from "./EventManager.js";
import { Request } from "./Request.js";
import { Response } from "./Response.js";

/**
 * A web socket client instance, a convenience front-end to the WebSocket type.
 *
 * @author <a href="mailto:jacob.dreyer@geosoft.no">Jacob Dreyer</a>
 */
export class WebSocketClient
{
  /** {string} Web socket URI. */
  #uri_;

  /** The low-level web socket instance. */
  #webSocket_;

  /** {Request[]} Requests made before the web socket was opened. */
  #requestQueue_ = [];

  /**
   * Create a web socket client instance.
   *
   * @param {string} uri - The web socket URI. Non-null.
   */
  constructor(uri)
  {
    if (uri == null)
      throw new TypeError("uri cannot be null");

    this.#uri_ = uri;
    this.#requestQueue_ = [];

    this.#webSocket_ = new WebSocket(uri);
    this.#webSocket_.onopen = this.onOpen.bind(this);
    this.#webSocket_.onclose = this.onClose.bind(this);
    this.#webSocket_.onmessage = this.onMessage.bind(this);
    this.#webSocket_.onerror = this.onError.bind(this);
  }

  /**
   * Called when the connection is opened.
   */
  onOpen()
  {
    console.log("WebSocket connection opened");

    EventManager.getInstance().notify("WebSocketConnectionOpened", this.#uri_)

    // Send any queued messages
    this.#requestQueue_.forEach(request => this.#webSocket_.send(request.toJson()));
    this.#requestQueue_ = []; // Clear the queue
  }

  /**
   * Called when the connection is closed.
   */
  onClose()
  {
    console.log("WebSocket connection closed");
    EventManager.getInstance().notify("WebSocketConnectionClosed", this.#uri_);
  }

  /**
   * Called when a message is received.
   */
  onMessage(event)
  {
    console.log("Received message: ", event.data);
    const response = new Response(event.data);
    EventManager.getInstance().notify("WebSocketResponseReceived", this.#uri_, response);
  }

  /**
   * Called when an error occurs.
   */
  onError(error)
  {
    console.log("WebSocket error: " + error);
  }

  /**
   * Check if the connection is open.
   *
   * @return {boolean}  True if the connection is open, false otherwise.
   */
  isOpen()
  {
    return this.#webSocket_.readyState == WebSocket.OPEN;
  }

  /**
   * Send a message through the web socket.
   *
   * @param {Request}  Request to send. Non-null.
   */
  send(request)
  {
    if (request == null || !(request instanceof Request))
      throw new TypeError("Invalid request: " + request);

    console.log("Send: " + request.toJson());

    // Send right away if the web socket is open
    if (this.isOpen()) {
      this.#webSocket_.send(request.toJson());
      EventManager.getInstance().notify("WebSocketRequestSent", this, request);
    }

    // Otherwise queue the message and send it as soon as the socket opens
    else {
      this.#requestQueue_.push(request);
    }
  }
}

File: java/src/no/geosoft/logiq/core/websocket/WebSocketClient.java
package no.geosoft.logiq.core.websocket;

import java.io.IOException;
import java.net.URI;
import java.util.logging.Level;
import java.util.logging.Logger;

import org.java_websocket.handshake.ServerHandshake;
import org.java_websocket.exceptions.WebsocketNotConnectedException;

import no.geosoft.cc.util.TextUtil;
import no.geosoft.cc.event.EventManager;

import no.geosoft.logiq.core.jsonrpc.Request;
import no.geosoft.logiq.core.jsonrpc.Response;

/**
 * A websocket client instance.
 *
 * @author <a href="mailto:jacob.dreyer@geosoft.no">Jacob Dreyer</a>
 */
public class WebSocketClient extends org.java_websocket.client.WebSocketClient
{
  /** The logger instance. */
  private static final Logger logger_ = Logger.getLogger(WebSocketClient.class.getName());

  /**
   * Create a websocket client.
   *
   * @param webSocketUri  URI to server web socket. Non-null.
   */
  public WebSocketClient(URI webSocketUri)
  {
    super(webSocketUri);
  }

  /** {@inheritDoc} */
  @Override
  public void onOpen(ServerHandshake serverHandshake)
  {
    logger_.log(Level.INFO, "Connection open " + getRemoteSocketAddress() +
                " - " + serverHandshake.getHttpStatusMessage() +
                " (" + serverHandshake.getHttpStatus() + ")");
    EventManager.getInstance().notify("LogIqConnectionOpened", this);
  }

  /** {@inheritDoc} */
  @Override
  public void onClose(int exitCode, String closeReason, boolean isRemote)
  {
    logger_.log(Level.INFO, "Connection closed " + getRemoteSocketAddress() +
                " - " + closeReason + " (" + exitCode + ")");
    EventManager.getInstance().notify("LogIqConnectionClosed", this);
  }

  /** {@inheritDoc} */
  @Override
  public void onMessage(String message)
  {
    logger_.log(Level.INFO, "Response: " + message);

    Response response = new Response(message);
    EventManager.getInstance().notify("LogIqResponseReceived", this, response);
  }

  /** {@inheritDoc} */
  @Override
  public void onError(Exception exception)
  {
    logger_.log(Level.WARNING, "Web socket error: " + getURI(), exception);
  }

  /**
   * Send the specified request to the server.
   *
   * @param message  Request to send. Non-null.
   * @throws IOException  If sending failed for some reason.
   */
  public void send(Request request)
    throws IOException
  {
    if (request == null)
      throw new IllegalArgumentException("request cannot be null");

    try {
      send(request.toJson());

      EventManager.getInstance().notify("LogIqRequestSent", this, request);
    }
    catch (WebsocketNotConnectedException exception) {
      throw new IOException("Unable to send message: " + TextUtil.truncate(request.toJson(), 40), exception);
    }
  }
}
